openraft/src/raft/mod.rs

```diff
@@ -780,10 +780,95 @@ where C: RaftTypeConfig
+    ///
+    /// The function `func` is applied to the current [`RaftStateMachine`]. The result of this
+    /// function, of type `V`, is returned wrapped in
+    /// `Result<Result<V, InvalidStateMachineType>, Fatal<C>>`.
+    /// `Fatal` error will be returned if failed to receive a reply from `RaftCore`.
+    ///
+    /// A `Fatal` error is returned if:
+    /// - Raft core task is stopped normally.
+    /// - Raft core task is panicked due to programming error.
+    /// - Raft core task is encountered a storage error.
+    ///
+    /// If the user function fail to run, e.g., the input `SM` is different one from the one in
+    /// `RaftCore`, it returns an [`InvalidStateMachineType`] error.
+    ///
+    /// Example for getting the last applied log id from SM(assume there is `last_applied()` method
+    /// provided):
+    ///
+    /// ```rust,ignore
+    /// let last_applied_log_id = my_raft.with_state_machine(|sm| {
+    ///     async move { sm.last_applied().await }
+    /// }).await?;
+    /// ```
+    pub async fn with_state_machine<F, SM, V>(&self, func: F) -> Result<Result<V, InvalidStateMachineType>, Fatal<C>>
+    where
+        SM: RaftStateMachine<C>,
+        F: FnOnce(&mut SM) -> BoxFuture<V> + OptionalSend + 'static,
+        V: OptionalSend + 'static,
+    {
+        let (tx, rx) = C::oneshot();
+
+        self.external_state_machine_request(|sm| {
+            Box::pin(async move {
+                let resp = func(sm).await;
+                if let Err(_err) = tx.send(resp) {
+                    tracing::error!("{}: fail to send response to user communicating tx", func_name!());
+                }
+            })
+        });
+
+        let recv_res = rx.await;
+        tracing::debug!("{} receives result is error: {:?}", func_name!(), recv_res.is_err());
+
+        let Ok(v) = recv_res else {
+            if self.inner.is_core_running().await {
+                return Ok(Err(InvalidStateMachineType::new::<SM>()));
+            } else {
+                let fatal = self.inner.get_core_stopped_error("receiving rx from RaftCore", None::<&'static str>).await;
+                tracing::error!(error = debug(&fatal), "error when {}", func_name!());
+                return Err(fatal);
+            }
+        };
+
+        Ok(Ok(v))
+    }
+
+    /// Send a request to the [`RaftStateMachine`] worker in a fire-and-forget manner.
+    ///
+    /// The request functor will be called with a mutable reference to the state machine.
+    /// The functor returns a [`Future`] because state machine methods are `async`.
+    ///
+    /// If the API channel is already closed (Raft is in shutdown), then the request functor is
+    /// destroyed right away and not called at all.
+    ///
+    /// If the input `SM` is different from the one in `RaftCore`, it just silently ignores it.
+    pub fn external_state_machine_request<F, SM>(&self, req: F)
+    where
+        SM: 'static,
```

Comment:
Noticed that `SM` does not have a `RaftStateMachine<C>` trait bound, is this something by design for greater flexibility? 

```rust
SM: 'static + RaftStateMachine<C>,
```

Though `SM` can indeed be anything as the underlying state machine worker would ignore incorrect type:

```rust
                    } else {
                        tracing::warn!(
                            "User-defined SM function uses incorrect state machine type, expected: {}, got: {}",
                            std::any::type_name::<SM>(),
                            input_sm_type
                        );
                    };
```

Reply:
It must be a mistake 🤔

Reply:
then let's fix it:)